File: MacDown/Code/Extension/NSTextView+Autocomplete.m
//
//  NSTextView+Autocomplete.m
//  MacDown
//
//  Created by Tzu-ping Chung  on 11/06/2014.
//  Copyright (c) 2014 Tzu-ping Chung . All rights reserved.
//

#import "NSTextView+Autocomplete.h"
#import "NSString+Lookup.h"
#import "MPUtilities.h"


static const unichar kMPMatchingCharactersMap[][2] = {
    {L'(', L')'},
    {L'[', L']'},
    {L'{', L'}'},
    {L'<', L'>'},
    {L'\'', L'\''},
    {L'\"', L'\"'},
    {L'\uff08', L'\uff09'},     // full-width parentheses
    {L'\u300c', L'\u300d'},     // corner brackets
    {L'\u300e', L'\u300f'},     // white corner brackets
    {L'\u2018', L'\u2019'},     // single quotes
    {L'\u201c', L'\u201d'},     // double quotes
    {L'\0', L'\0'},
};

static const unichar kMPStrikethroughCharacter = L'~';

static const unichar kMPMarkupCharacters[] = {
    L'*', L'_', L'`', L'=', L'\0',
};

static NSString * const kMPListLineHeadPattern =
    @"^(\\s*)((?:(?:\\*|\\+|-|)\\s+)?)((?:\\d+\\.\\s+)?)(\\S)?";


@implementation NSTextView (Autocomplete)

- (BOOL)substringInRange:(NSRange)range isSurroundedByPrefix:(NSString *)prefix
                  suffix:(NSString *)suffix
{
    NSString *content = self.string;
    NSUInteger location = range.location;
    NSUInteger length = range.length;
    if (content.length < location + length + suffix.length)
        return NO;
    if (location < prefix.length)
        return NO;

    if (![[content substringFromIndex:location + length] hasPrefix:suffix]
        || ![[content substringToIndex:location] hasSuffix:prefix])
        return NO;

    // Emphasis (*) requires special treatment because we need to eliminate
    // strong (**) but not strong-emphasis (***).
    if (![prefix isEqualToString:@"*"] || ![suffix isEqualToString:@"*"])
        return YES;
    if ([self substringInRange:range isSurroundedByPrefix:@"***" suffix:@"***"])
        return YES;
    if ([self substringInRange:range isSurroundedByPrefix:@"**" suffix:@"**"])
        return NO;
    return YES;
}


- (void)insertSpacesForTab
{
    NSString *spaces = @"    ";
    NSUInteger currentLocation = self.selectedRange.location;
    NSInteger p = [self.string locationOfFirstNewlineBefore:currentLocation];

    // Calculate how deep we need to go.
    NSUInteger offset = (currentLocation - p - 1) % 4;
    if (offset)
        spaces = [spaces substringFromIndex:offset];
    [self insertText:spaces];
}

- (BOOL)completeMatchingCharactersForTextInRange:(NSRange)range
                                      withString:(NSString *)str
                            strikethroughEnabled:(BOOL)strikethrough
{
    NSUInteger stringLength = str.length;

    // Character insert without selection.
    if (range.length == 0 && stringLength == 1)
    {
        NSUInteger location = range.location;
        if ([self completeMatchingCharacterForText:str
                                        atLocation:location])
            return YES;
    }
    // Character insert with selection (i.e. select and replace).
    else if (range.length > 0 && stringLength == 1)
    {
        unichar character = [str characterAtIndex:0];
        if ([self wrapMatchingCharactersOfCharacter:character
                                  aroundTextInRange:range
                               strikethroughEnabled:strikethrough])
            return YES;
    }
    return NO;
}

- (BOOL)completeMatchingCharacterForText:(NSString *)string
                              atLocation:(NSUInteger)location
{
    NSString *content = self.string;
    NSUInteger contentLength = content.length;

    unichar c = [string characterAtIndex:0];
    unichar n = ' ';
    unichar p = ' ';
    if (location < contentLength)
        n = [content characterAtIndex:location];
    if (location > 0 && location <= contentLength)
        p = [content characterAtIndex:location - 1];

    NSCharacterSet *delims = [NSCharacterSet whitespaceAndNewlineCharacterSet];
    for (const unichar *cs = kMPMatchingCharactersMap[0]; *cs != 0; cs += 2)
    {
        if ([delims characterIsMember:n] && c == cs[0] && n != cs[1]
            && ([delims characterIsMember:p] || cs[0] != cs[1]))
        {
            NSRange range = NSMakeRange(location, 0);
            NSString *completion = [NSString stringWithCharacters:cs length:2];
            [self insertText:completion replacementRange:range];

            range.location += string.length;
            self.selectedRange = range;
            return YES;
        }
        else if (c == cs[1] && n == cs[1])
        {
            NSRange range = NSMakeRange(location + 1, 0);
            self.selectedRange = range;
            return YES;
        }
    }
    return NO;
}

- (void)wrapTextInRange:(NSRange)range withPrefix:(unichar)prefix
                 suffix:(unichar)suffix
{
    NSString *string = [self.string substringWithRange:range];
    NSString *p = [NSString stringWithCharacters:&prefix length:1];
    NSString *s = [NSString stringWithCharacters:&suffix length:1];
    NSString *wrapped = [NSString stringWithFormat:@"%@%@%@", p, string, s];
    [self insertText:wrapped replacementRange:range];

    range.location += 1;
    self.selectedRange = range;
}

- (BOOL)wrapMatchingCharactersOfCharacter:(unichar)character
                        aroundTextInRange:(NSRange)range
                     strikethroughEnabled:(BOOL)isStrikethroughEnabled
{
    for (const unichar *cs = kMPMatchingCharactersMap[0]; *cs != 0; cs += 2)
    {
        if (character == cs[0])
        {
            [self wrapTextInRange:range withPrefix:cs[0] suffix:cs[1]];
            return YES;
        }
    }
    for (size_t i = 0; kMPMarkupCharacters[i] != 0; i++)
    {
        if (character == kMPMarkupCharacters[i])
        {
            [self wrapTextInRange:range withPrefix:character suffix:character];
            return YES;
        }
    }
    if (isStrikethroughEnabled && character == kMPStrikethroughCharacter)
    {
        [self wrapTextInRange:range withPrefix:character suffix:character];
        return YES;
    }
    return NO;
}

- (void)deleteMatchingCharactersAround:(NSUInteger)location
{
    NSString *string = self.string;
    if (location == 0 || location >= string.length)
        return;

    unichar f = [string characterAtIndex:location - 1];
    unichar b = [string characterAtIndex:location];

    for (const unichar *cs = kMPMatchingCharactersMap[0]; *cs != 0; cs += 2)
    {
        if (f == cs[0] && b == cs[1])
        {
            [self replaceCharactersInRange:NSMakeRange(location, 1)
                                withString:@""];
            break;
        }
    }
}

- (void)unindentForSpacesBefore:(NSUInteger)location
{
    NSString *string = self.string;

    NSUInteger whitespaceCount = 0;
    while (location - whitespaceCount > 0
           && [string characterAtIndex:location - whitespaceCount - 1] == L' ')
    {
        whitespaceCount++;
        if (whitespaceCount >= 4)
            break;
    }
    if (whitespaceCount < 2)
        return;

    NSUInteger offset =
        ([self.string locationOfFirstNewlineBefore:location] + 1) % 4;
    if (offset == 0)
        offset = 4;
    offset = offset > whitespaceCount ? whitespaceCount : 4;
    NSRange range = NSMakeRange(location - offset, offset);

    // Leave a space for the original delete action to handle.
    [self replaceCharactersInRange:range withString:@" "];
}

- (BOOL)toggleForMarkupPrefix:(NSString *)prefix suffix:(NSString *)suffix
{
    NSRange range = self.selectedRange;
    NSString *selection = [self.string substringWithRange:range];
    BOOL isOn = NO;

    // Selection is already marked-up. Clear markup and maintain selection.
    NSUInteger poff = prefix.length;
    if ([self substringInRange:range isSurroundedByPrefix:prefix
                        suffix:suffix])
    {
        NSRange sub = NSMakeRange(range.location - poff,
                                  selection.length + poff + suffix.length);
        [self insertText:selection replacementRange:sub];
        range.location = sub.location;
        isOn = NO;
    }
    // Selection is normal. Mark it up and maintain selection.
    else
    {
        NSString *text = [NSString stringWithFormat:@"%@%@%@",
                          prefix, selection, suffix];
        [self insertText:text replacementRange:range];
        range.location += poff;
        isOn = YES;
    }
    self.selectedRange = range;
    return isOn;
}

- (void)toggleBlockWithPattern:(NSString *)pattern prefix:(NSString *)prefix
{
    NSRegularExpression *regex =
        [[NSRegularExpression alloc] initWithPattern:pattern options:0
                                               error:NULL];
    NSString *content = self.string;
    NSRange selectedRange = self.selectedRange;
    NSRange lineRange = [content lineRangeForRange:selectedRange];

    NSString *toProcess = [content substringWithRange:lineRange];
    NSArray *lines = [toProcess componentsSeparatedByString:@"\n"];

    BOOL isMarked = YES;
    for (NSString *line in lines)
    {
        NSUInteger lineLength = line.length;
        if (!lineLength)
            continue;
        NSRange matchRange =
            [regex rangeOfFirstMatchInString:line options:0
                                       range:NSMakeRange(0, lineLength)];
        if (matchRange.location == NSNotFound)
        {
            isMarked = NO;
            break;
        }
    }

    NSUInteger prefixLength = prefix.length;
    NSMutableArray *modLines = [NSMutableArray arrayWithCapacity:lines.count];

    __block NSUInteger totalShift = 0;
    [lines enumerateObjectsUsingBlock:^(id obj, NSUInteger index, BOOL *stop) {
        NSString *line = obj;
        if (line.length)
        {
            totalShift += prefixLength;
            if (!isMarked)
                line = [prefix stringByAppendingString:line];
            else
                line = [line substringFromIndex:prefixLength];
        }
        [modLines addObject:line];
    }];
    NSString *processed = [modLines componentsJoinedByString:@"\n"];
    [self insertText:processed replacementRange:lineRange];

    if (!isMarked)
    {
        selectedRange.location += prefixLength;
        selectedRange.length += totalShift - prefixLength;
    }
    else
    {
        selectedRange.location -= prefixLength;
        selectedRange.length -= totalShift - prefixLength;
        if (selectedRange.location < lineRange.location)
        {
            selectedRange.length -= lineRange.location - selectedRange.location;
            selectedRange.location = lineRange.location;
        }
    }
    self.selectedRange = selectedRange;
}

- (void)indentSelectedLinesWithPadding:(NSString *)padding
{
    NSString *content = self.string;
    NSRange selectedRange = self.selectedRange;
    NSRange lineRange = [content lineRangeForRange:selectedRange];

    NSString *toProcess = [content substringWithRange:lineRange];
    NSArray *lines = [toProcess componentsSeparatedByString:@"\n"];
    NSMutableArray *modLines = [NSMutableArray arrayWithCapacity:lines.count];
    NSUInteger paddingLength = padding.length;

    __block NSUInteger totalShift = 0;
    [lines enumerateObjectsUsingBlock:^(id obj, NSUInteger index, BOOL *stop) {
        NSString *line = obj;
        if (line.length)
        {
            totalShift += paddingLength;
            line = [padding stringByAppendingString:line];
        }
        [modLines addObject:line];
    }];
    NSString *processed = [modLines componentsJoinedByString:@"\n"];
    [self insertText:processed replacementRange:lineRange];

    selectedRange.location += paddingLength;
    selectedRange.length += totalShift - paddingLength;
    self.selectedRange = selectedRange;
}

- (void)unindentSelectedLines
{
    NSString *content = self.string;
    NSRange selectedRange = self.selectedRange;
    NSRange lineRange = [content lineRangeForRange:selectedRange];

    NSString *toProcess = [content substringWithRange:lineRange];
    NSArray *lines = [toProcess componentsSeparatedByString:@"\n"];
    NSMutableArray *modLines = [NSMutableArray arrayWithCapacity:lines.count];

    __block NSUInteger firstShift = 0;
    __block NSUInteger totalShift = 0;
    [lines enumerateObjectsUsingBlock:^(id obj, NSUInteger index, BOOL *stop) {
        NSString *line = obj;
        NSUInteger lineLength = line.length;
        NSUInteger shift = 0;
        for (shift = 0; shift <= 4; shift++)
        {
            if (shift >= lineLength)
                break;
            unichar c = [line characterAtIndex:shift];
            if (c == '\t')
                shift++;
            if (c != ' ')
                break;
        }
        if (index == 0)
            firstShift += shift;
        totalShift += shift;
        if (shift && shift < lineLength)
            line = [line substringFromIndex:shift];
        [modLines addObject:line];
    }];
    NSString *processed = [modLines componentsJoinedByString:@"\n"];
    [self insertText:processed replacementRange:lineRange];

    selectedRange.location -= firstShift;
    selectedRange.length -= totalShift - firstShift;
    self.selectedRange = selectedRange;
}

- (BOOL)insertMappedContent
{
    NSString *content = self.string;
    NSUInteger contentLength = content.length;
    if (contentLength > 20)
        return NO;
    static NSDictionary *map = nil;
    if (!map)
    {
        NSBundle *bundle = [NSBundle mainBundle];
        NSString *filePath = [bundle pathForResource:@"data" ofType:@"map"
                                         inDirectory:@"data"];
        map = [NSKeyedUnarchiver unarchiveObjectWithFile:filePath];
    }
    NSData *mapped = map[content];
    if (!mapped)
        return NO;
    NSArray *components = @[NSTemporaryDirectory(),
                             [NSString stringWithFormat:@"%lu", content.hash]];
    NSString *path = [NSString pathWithComponents:components];
    [mapped writeToFile:path atomically:NO];
    NSString *text = [NSString stringWithFormat:@"![%@](%@)", content, path];
    [self insertText:text replacementRange:NSMakeRange(0, contentLength)];
    self.selectedRange = NSMakeRange(2, contentLength);
    return YES;
}


- (BOOL)completeNextLine
{
    NSRange selectedRange = self.selectedRange;
    NSUInteger location = selectedRange.location;
    NSString *content = self.string;
    if (selectedRange.length || !content.length)
        return NO;

    NSInteger start = [content locationOfFirstNewlineBefore:location] + 1;
    NSUInteger end = location;
    NSUInteger nonwhitespace =
        [content locationOfFirstNonWhitespaceCharacterInLineBefore:location];

    // No non-whitespace character at this line.
    if (nonwhitespace == location)
        return NO;

    NSRange range = NSMakeRange(start, end - start);
    NSString *line = [self.string substringWithRange:range];


    NSRegularExpressionOptions options = NSRegularExpressionAnchorsMatchLines;
    NSRegularExpression *regex =
        [[NSRegularExpression alloc] initWithPattern:kMPListLineHeadPattern
                                             options:options
                                               error:NULL];
    NSTextCheckingResult *result =
        [regex firstMatchInString:line options:0
                            range:NSMakeRange(0, line.length)];
    if (!result || result.range.location == NSNotFound)
        return NO;

    NSMutableString *indent = [[NSMutableString alloc] init];
    for (NSUInteger i = 0; i < [result rangeAtIndex:1].length; i++)
        [indent appendString:@" "];

    NSString *t = nil;
    BOOL isUl = ([result rangeAtIndex:2].length != 0);
    BOOL isOl = ([result rangeAtIndex:3].length != 0);
    BOOL previousLineEmpty = ([result rangeAtIndex:4].length == 0);
    if (previousLineEmpty)
    {
        NSRange replaceRange = NSMakeRange(NSNotFound, 0);
        if (isUl)
            replaceRange = [result rangeAtIndex:2];
        else if (isOl)
            replaceRange = [result rangeAtIndex:3];
        if (replaceRange.length)
        {
            replaceRange.location += start;
            [self replaceCharactersInRange:range withString:@""];
        }
    }
    else if (isUl)
    {
        NSRange range = [result rangeAtIndex:2];
        range.length -= 1;      // Exclude trailing whitespace.
        t = [line substringWithRange:range];
    }
    else if (isOl)
    {
        NSRange range = [result rangeAtIndex:3];
        range.length -= 1;      // Exclude trailing space.
        NSString *captured = [line substringWithRange:range];
        NSInteger i = captured.integerValue + 1;
        t = [NSString stringWithFormat:@"%ld.", i];
    }
    [self insertNewline:self];
    if (!t)
        return YES;

    location += 1;  // Shift for inserted newline.
    NSString *it = [NSString stringWithFormat:@"%@%@", indent, t];
    NSUInteger contentLength = content.length;

    // Has matching list item. Only insert indent.
    NSRange r = NSMakeRange(location, t.length);
    if (contentLength > location + t.length
            && [[content substringWithRange:r] isEqualToString:t])
    {
        [self insertText:indent];
        return YES;
    }

    // Has indent and matching list item. Accept it.
    r = NSMakeRange(location, it.length);
    if (contentLength > location + it.length
            && [[content substringWithRange:r] isEqualToString:it])
        return YES;

    // Insert completion for normal cases.
    [self insertText:[NSString stringWithFormat:@"%@ ", it]];
    return YES;
}

@end

File: MacDown/Code/Extension/NSTextView+Autocomplete.swift
//
//  NSTextView+Autocomplete.swift
//  MacDown
//
//  Created by Foster Yin on 7/7/14.
//  Copyright (c) 2014 MacUp, MarkDown, Swift. All rights reserved.
//

import Cocoa

let kMUSMatchingCharactersMap:Character[][] = [
    ["(", ")"],
    ["[", "]"],
    ["[", "]"],
    ["<", ">"],
    ["\"", "\""],
    ["\"", "\""],
    ["\uff08", "\uff09"],     // full-width parentheses
    ["\u300c", "\u300d"],     // corner brackets
    ["\u300e", "\u300f"],     // white corner brackets
    ["\u2018", "\u2019"],     // single quotes
    ["\u201c", "\u201d"],     // double quotes
    ["\0", "\0"],]

let kMUSStrikethroughCharacter:Character = "~"

let kMUSMarkupCharacters:Character[] = [
    "*", "_", "`", "=", "\0",]

let kMUSListLineHeadPattern = "^(\\s*)((?:(?:\\*|\\+|-|)\\s+)?)((?:\\d+\\.\\s+)?)(\\S)?"

let kMUSDataMap:Dictionary<String, NSData> = {
    let bundle = NSBundle.mainBundle()
    let filePath = bundle.pathForResource("data", ofType: "map", inDirectory:"data")
    let map : AnyObject! = NSKeyedUnarchiver.unarchiveObjectWithFile(filePath)
    return map as Dictionary<String, NSData>
}()

extension NSTextView
{
    func substringInRange(range:NSRange, isSurroundedByPrefix prefix:String?, suffix:String?) -> Bool
    {
        let content = self.string

        let location = range.location
        let length = range.length

        if (content.lengthOfBytesUsingEncoding(NSUTF8StringEncoding) < location + length + suffix!.lengthOfBytesUsingEncoding(NSUTF8StringEncoding))
        {
            return false
        }

        if (location < prefix!.lengthOfBytesUsingEncoding(NSUTF8StringEncoding))
        {
            return false
        }

        if (!content.substringFromIndex(location + length).hasPrefix(suffix!) ||
            !content.substringToIndex(location).hasSuffix(prefix!))
        {
            return false
        }

        if (!(prefix! == "*") || !(suffix! == "*"))
        {
            return true
        }
        if (self.substringInRange(range, isSurroundedByPrefix: "***", suffix: "***"))
        {
            return true
        }
        if (self.substringInRange(range, isSurroundedByPrefix: "**", suffix: "**"))
        {
            return false
        }

        return true

    }

    func insertSpacesForTab()
    {
        var spaces = "    "
        let currentLocation = self.selectedRange.location
        let p = self.string.locationOfFirstNewlineBefore(currentLocation)
        let offset = (currentLocation - p - 1) % 4
        if (offset > 0)
        {
            spaces = spaces.substringFromIndex(offset)
        }

        self.insertText(spaces)
    }

    func completeMatchingCharactersForTextInRange(range:NSRange, withString string:String?, strikethroughEnabled:Bool) -> Bool
    {
        let stringLength = string!.lengthOfBytesUsingEncoding(NSUTF8StringEncoding)

        if (range.length == 0 && stringLength == 1)
        {
            let location = range.location
            if (self.completeMatchingCharacterForText(string, atLocation: location))
            {
                return true
            }
        }
        else if (range.length > 0 && stringLength == 1)
        {
            let char:Character = string![0]
            if (self.wrapMatchingCharactersOfCharacter(char, aroundTextInRange: range, strikethroughEnabled: strikethroughEnabled))
            {
                return true
            }
        }
        
        return false
    }


    func completeMatchingCharacterForText(string:String?, atLocation location:Int) -> Bool
    {
        let content = self.string
        let contentLength = content.lengthOfBytesUsingEncoding(NSUTF8StringEncoding)

        let c:Character = string![0]
        var n:Character = " "
        var p:Character = " "

        if (location < contentLength)
        {
            n = content[location]
        }
        if (location > 0 && location <= contentLength)
        {
            p = content[location - 1]
        }

        let delims = NSCharacterSet.whitespaceAndNewlineCharacterSet()

        for cs:Character[] in kMUSMatchingCharactersMap
        {
            if (delims.characterIsMember(n.toUnichar())
                && (c == cs[0])
                && (n != cs[1])
                && (delims.characterIsMember(p.toUnichar()) || "\(cs[0])" != "\(cs[1])"))
            {
                let range = NSMakeRange(location, 0)
                let completion = "" + cs[0] + cs[1]
                self.insertText(completion, replacementRange: range)

                self.selectedRange = NSMakeRange(range.location, range.location + string!.lengthOfBytesUsingEncoding(NSUTF8StringEncoding))
                return true
            }
            else if (c == cs[1] && n == cs[1])
            {
                let range = NSMakeRange(location + 1, 0)
                self.selectedRange = range
                return true
            }
        }

        return false
    }

    func wrapTextInRange(range:NSRange, withPrefix prefix:Character, suffix:Character)
    {
        let string = self.string.substring(range.location, length: range.length)
        let wrapped = prefix + string + suffix
        self.insertText(wrapped, replacementRange: range)
        self.selectedRange = NSMakeRange(range.location + 1, range.length)
    }

    func wrapMatchingCharactersOfCharacter(character:Character, aroundTextInRange range:NSRange, strikethroughEnabled:Bool) -> Bool
    {
        for cs:Character[] in kMUSMatchingCharactersMap
        {
            if (character == cs[0])
            {
                self.wrapTextInRange(range, withPrefix: cs[0], suffix: cs[1])
                return true
            }
        }

        for char in kMUSMarkupCharacters
        {
            if (character == char)
            {
                self.wrapTextInRange(range, withPrefix: character, suffix: character)
                return true
            }
        }

        if (strikethroughEnabled && character == kMUSStrikethroughCharacter)
        {
            self.wrapTextInRange(range, withPrefix: character, suffix: character)
            return true
        }

        return false
    }

    func deleteMatchingCharactersAround(location:Int)
    {
        let string = self.string
        if (location == 0 || location == string.lengthOfBytesUsingEncoding(NSUTF8StringEncoding))
        {
            return
        }

        let f:Character = string[location - 1]
        let b:Character = string[location]

        for cs:Character[] in kMUSMatchingCharactersMap
        {
            if (f == cs[0] && b == cs[1])
            {
                self.replaceCharactersInRange(NSMakeRange(location, 1), withString: "")
                break
            }
        }
    }

    func unindentForSpacesBefore(location:Int)
    {
        let string = self.string

        var whitespaceCount = 0

        while ((location - whitespaceCount > 0)
            && (string[location - whitespaceCount - 1] == " "))
        {
            whitespaceCount++
            if (whitespaceCount >= 4)
            {
                break
            }
        }

        if (whitespaceCount < 2)
        {
            return
        }

        var offset = (self.string.locationOfFirstNewlineBefore(location) + 1) % 4
        if (offset == 0)
        {
            offset = 4
        }

        offset = offset > whitespaceCount ? whitespaceCount : 4
        let range = NSMakeRange(location - offset, offset)
        self.replaceCharactersInRange(range, withString: " ")
    }

    func toggleForMarkupPrefix(prefix:String, suffix:String) -> Bool
    {
        var range = self.selectedRange

        let selection = self.string.substring(range.location, length: range.length)

        var isOn = false

        let poff = prefix.lengthOfBytesUsingEncoding(NSUTF8StringEncoding)

        if (self.substringInRange(range, isSurroundedByPrefix: prefix, suffix: suffix))
        {
            let sub = NSMakeRange(range.location - poff, selection.lengthOfBytesUsingEncoding(NSUTF8StringEncoding) + poff + suffix.lengthOfBytesUsingEncoding(NSUTF8StringEncoding))
            self.insertText(selection, replacementRange: sub)
            range.location = sub.location
            isOn = false
        }
        else
        {
            let text = prefix + selection + suffix
            self.insertText(text, replacementRange: range)
            range.location += poff
            isOn = true
        }

        self.selectedRange = range
        return isOn
    }

    func toggleBlockWithPattern(pattern:String, prefix:String)
    {
        let regex = NSRegularExpression(pattern: pattern, options:NSRegularExpressionOptions(0) , error: nil)

        let content = self.string

        var selectedRange = self.selectedRange

        let lineRange = content.lineRangeForRange(content.NSRangeToRange(selectedRange))

        let toProcess = content.substringWithRange(lineRange)
        let lines:Array<String> = toProcess.componentsSeparatedByString("\n")

        var isMarked = true

        for line in lines
        {
            let lineLength = line.lengthOfBytesUsingEncoding(NSUTF8StringEncoding)
            if (lineLength == 0)
            {
                continue
            }

            let matchRange = regex.rangeOfFirstMatchInString(line, options: NSMatchingOptions(0), range: NSMakeRange(0, lineLength))

            if (matchRange.location == NSNotFound)
            {
                isMarked = false
                break
            }
        }

        let prefixLength = prefix.lengthOfBytesUsingEncoding(NSUTF8StringEncoding)

        var modLines = Array<String>()


        var totalShift = 0
        for line in lines
        {
            var theLine = line
            if (line.lengthOfUTF8() > 0)
            {
                totalShift += prefixLength

                if (!isMarked)
                {
                    theLine = prefix + line
                }
                else
                {
                    theLine = line.substringFromIndex(prefixLength)
                }
            }
            modLines.append(theLine)
        }

        let processed = (modLines as NSArray).componentsJoinedByString("\n")

        self.insertText(processed, replacementRange: content.RangeToNSRange(lineRange))

        if (!isMarked)
        {
            selectedRange.location += prefixLength
            selectedRange.length += (totalShift - prefixLength)
        }
        else
        {
            selectedRange.location -= prefixLength
            selectedRange.length -= (totalShift - prefixLength)

            if (selectedRange.location < content.RangeToNSRange(lineRange).location)
            {
                selectedRange.length -= (content.RangeToNSRange(lineRange).location - selectedRange.location)
                selectedRange.location = content.RangeToNSRange(lineRange).location
            }
        }

        self.selectedRange = selectedRange

    }

    func indentSelectedLinesWithPadding(padding:String)
    {
        let content = self.string
        var selectedRange = self.selectedRange
        let lineRange = content.lineRangeForRange(content.NSRangeToRange(selectedRange))

        let toProcess = content.substringWithRange(lineRange)
        let lines:Array<String> = toProcess.componentsSeparatedByString("\n")

        var modLines = Array<String>()
        let paddingLength = padding.lengthOfUTF8()

        var totalShift = 0

        for line in lines
        {
            var theLine = line
            if (line.lengthOfUTF8() > 0)
            {
                totalShift += paddingLength
                theLine = padding.stringByAppendingString(line)
            }

            modLines += theLine
        }

        let processed = (modLines as NSArray).componentsJoinedByString("\n")
        self.insertText(processed, replacementRange: content.RangeToNSRange(lineRange))

        selectedRange.location += paddingLength
        selectedRange.length += (totalShift - paddingLength)
        self.selectedRange = selectedRange
    }

    func unindentSelectedLines()
    {
        let content = self.string
        var selectedRange = self.selectedRange
        let lineRange = content.lineRangeForRange(content.NSRangeToRange(selectedRange))

        let toProcess = content.substringWithRange(lineRange)
        let lines:Array<String> = toProcess.componentsSeparatedByString("\n")

        var modLines = Array<String>()

        var firstShift = 0
        var totalShift = 0
        for (index, line) in enumerate(lines)
        {
            let lineLength = line.lengthOfUTF8()
            var theLine = line
            var shift = 0
            for (shift = 0; shift <= 4; shift++)
            {
                if (shift >= lineLength)
                {
                    break
                }

                let char = line.substring(shift, length: 1)
                if (char == "\t")
                {
                    shift++
                }
                if (char != " ")
                {
                    break
                }

                if (index == 0)
                {
                    firstShift += shift
                }

                totalShift += shift

                if ((shift > 0) && (shift < lineLength))
                {
                    theLine = line.substringFromIndex(shift)
                }
                modLines += theLine
            }
        }

        let processed = (modLines as NSArray).componentsJoinedByString("\n")
        self.insertText(processed, replacementRange: content.RangeToNSRange(lineRange))

        selectedRange.location -= firstShift
        selectedRange.length -= (totalShift - firstShift)
        self.selectedRange = selectedRange

    }


    func insertMappedContent() -> Bool
    {
        let content = self.string
        let contentLength = content.lengthOfUTF8()
        if (contentLength > 20)
        {
            return false

        }

        let mapped = kMUSDataMap[content]
        if (!mapped)
        {
            return false
        }

        let path = NSTemporaryDirectory() + "/" + "\(content.hash)"
        mapped!.writeToFile(path, atomically: false)
        let text = "![\(content)](\(path))"
        self.insertText(text, replacementRange: NSMakeRange(0, contentLength))
        self.selectedRange = NSMakeRange(2, contentLength)
        return true
    }
}
